Add docs to run Velero locally

carlisia opened the pull request:
I thought this would be useful. This is the best place I thought for it but there might be a better place. Documentation for how to run Velero with a local cluster (Kind, etc) could be added to this same page later.

Note: I ran the site locally and was getting the new link on the left navigation with same pages but not other (regardless of the version selected). Not sure if it's a local cache issue or if I missed something.

carlisia commented:
Nvm about my note, I checked the Netlify preview and it all looks right.

nrb commented on site/docs/v0.11.0/run-locally.md:
```diff
@@ -0,0 +1,25 @@
+# Run Velero locally
```
I'm not sure we want this in v0.11.0 - I'd consider anything prior to 1.0 "legacy" and not really requiring updates.

prydonius commented on site/_data/v1-1-0-toc.yml:
```diff
@@ -25,6 +25,8 @@ toc:
         url: /gcp-config
       - page: Restic setup
         url: /restic
+      - page: Run locally
```
The page should probably be called "run locally in development" since we tell users to build from the master branch

carlisia replied:
How about I add that at the top of the page itself, just to avoid having this menu item be so long?

skriss replied:
I'd prefer to not include this in the `Install` section, given that it's mainly a development topic, and it's in the `Contribute` section already.

skriss replied:
This is the last thing for me, then LGTM!

carlisia replied:
I thought I had removed it. Removing!

prydonius commented on site/docs/master/run-locally.md:
```diff
@@ -0,0 +1,25 @@
+
+Note: Since you are running locally, chances are you will want to run with the log level at debug, so add that flag.
+
+`./_output/bin/darwin/amd64/velero server --log-level debug`
```
isn't an env var needed for aws/azure/google? e.g. `AWS_SHARED_CREDENTIALS_FILE`

carlisia replied:
Is it? I ran w/o it and it "worked". @nrb pls confirm.

nrb replied:
It depends on the provider - each provider we have in-tree uses a credential "chain" that looks up the necessary auth information in env vars, config files, etc. It's totally possibly there was a local profile file that it picked up.

In general, if you look at the Velero `Deployment`, we use env vars to provide the Kubernetes `Secret` file to the server process. Because of that, I use the appropriate env var for the provider, to more closely match how the pod works. eg. `GOOGLE_APPLICATION_CREDENTIALS=bogus_file $VELERO/_output/bin/linux/amd64/velero server --log-level debug` (`$VELERO` is a local env var I use as a shortcut to my Velero checkout)

skriss commented:
We already have a docs section on this here: https://velero.io/docs/v1.1.0/build-from-source/. Happy to enhance/reorganize that page but let's not duplicate this info.  

Right now the "build from source" page is linked to from the "Customize" section in the TOC, which is a mix of contributor and user information. I'd propose moving the user-facing pages (run in any namespace, hooks, *maybe* the plugins page) into the "Use" section, and changing "Customize" to "Contribute".  Then we can keep developer-focused docs like this in that section.

carlisia commented:
Note: going to keep making changes only to the `master` tag until it's all reviewed.

Intention of changes: separate info on `build`, `run`, and `code`.

carlisia commented:
More specific notes:

- removed TOC from the build page since page (as well as other new pages) is now much shorter
- deleted a bunch of link references that were not being used
- batched info on running `make` under the same section
- deleted info on how to run as deployment, since that info is under the Install section

Question: imo, in the build/make section, the `make local` is adding nothing but confusion. Should we scratch that?

prydonius commented on site/_data/master-toc.yml:
```diff
@@ -45,18 +53,14 @@ toc:
-        url: /hooks
+      - page: Run locally
+        url: /run-locally
+      - page: Add code
```
The title of this page is a little confusing to me. I wonder if this should be called "Developer guide" or something else, as the content seems to be general development information.

prydonius commented on site/docs/master/build-from-source.md:
```diff
@@ -29,73 +22,42 @@ Where `go` is your [import path][4] for Go.
-The latter will place the compiled binary under `$PWD/_output/bin/$GOOS/$GOARCH`, and will splice version and git commit information in so that `velero version` displays proper output. `velero install` will also use the version information to determine which tagged image to deploy.
-
-To build the `velero` binary targeting `linux/amd64` within a build container on your local machine, run:
+This builds binaries for both linux and darwin. You will find the binary for darwing here: `_output/bin/darwin/amd64/velero`, and the binary for linux here: `_output/bin/linux/amd64/velero`.
```
```suggestion
This builds binaries for both linux and darwin. You will find the binary for darwin here: `_output/bin/darwin/amd64/velero`, and the binary for linux here: `_output/bin/linux/amd64/velero`.
```

prydonius commented on site/docs/master/build-from-source.md:
```diff
@@ -105,146 +67,20 @@ Velero's `Makefile` has a convenience target, `all-build`, that builds the follo
-This examples assumes you are using an existing cluster in AWS.
-
-Using the `velero` binary that you've built, run `velero install`:
+To build a Velero container image, first set the `$REGISTRY` environment variable. For example, if you want to build the `gcr.io/my-registry/velero:master` image, set `$REGISTRY` to `gcr.io/my-registry`. Optionally, set the `$VERSION` environment variable to change the image tag. Then, run:
```
We may want to mention that the default is gcr.io/heptio-images

prydonius commented on site/docs/master/run-locally.md:
```diff
@@ -0,0 +1,110 @@
+* A [BackupStorageLocation][20] object definition for the Velero server
+* (Optional) A [VolumeSnapshotLocation][21] object definition for the Velero server, to take PV snapshots
+
+### 1. Create a cluster
```
This makes it sound like you need a cluster on AWS to run velero locally, perhaps we should just link to the Install section which already goes through various cluster options on different platforms (including local) instead of pointing out specific options here?

prydonius commented on site/docs/master/run-locally.md:
```diff
@@ -0,0 +1,110 @@
+
+GCP: [GOOGLE_APPLICATION_CREDENTIALS][16]
+
+Azure:
```
on Azure you can actually set `AZURE_CREDENTIALS_FILE` to the file created at the bottom of this section: https://velero.io/docs/v1.1.0/azure-config/#create-service-principal

prydonius commented on site/docs/master/run-locally.md:
```diff
@@ -0,0 +1,110 @@
+
+Scale the Velero deployment down to 0 so it is not simultaneously being run on the remote cluster and potentially causing things to get out of sync:
+
+`kubectl scale --replicas=0 deployment velero -n velero`
```
In the previous section, the deployment is deleted. IMO we should just describe scaling it to 0, that way it's easily recoverable.

carlisia commented:
Note: I don't want to add to this PR, but these two issues would be additions to this new `Contribute` section:

- https://github.com/heptio/velero/issues/1387
- https://github.com/heptio/velero/issues/1573

skriss commented on site/docs/master/README.md:
```diff
@@ -36,19 +36,7 @@ If you encounter issues, review the [troubleshooting docs][30], [file an issue][
-### Pull requests
-
-* We welcome pull requests. Feel free to dig through the [issues][4] and jump in.
+If you are ready to jump in and test, add code, or help with documentation, follow the instructions on our [Start contributing](https://velero.io/docs/master/start-contributing/) documentation for guidance on how to setup Velero for development.
```
We probably want to make this change to the root repo `README.md` as well since this is just a copy of it

skriss commented on README.md:
```diff
@@ -70,11 +56,10 @@ See [the list of releases][6] to find out about feature changes.
-[28]: https://velero.io/docs/install-overview
-[29]: https://velero.io/docs/
-[30]: https://velero.io/docs/troubleshooting
+[28]: install-overview.md
```
Need to revert these link changes (28-100) - they still need to point to `velero.io/...`

skriss commented:
FYI https://github.com/heptio/velero/pull/1852#discussion_r324790494
